aws/data_source_aws_cloudhsm2_cluster.go

```diff
@@ -0,0 +1,128 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudhsmv2"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceCloudHsm2Cluster() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceCloudHsm2ClusterRead,
+
+		Schema: map[string]*schema.Schema{
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"cluster_state": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cluster_certificates": {
+				Type:     schema.TypeSet,
```

Comment:
I thing TypeSet is better in this context, because the number of available cluster certificates vary depending on the state of the cluster. For uninitialized cluster there's no certificates. Adding at least one HSM module to uninitialized cluster makes CSR and 3 other hardware certificates available, but as soon as cluster is initialized and activated there's the only one cluster certificate.
I might be wrong but I believe TypeList is a bit misleading in this context.

Reply:
It can be a little confusing, but the `Type` here is only referring to the parent attribute (`cluster_certificates`) itself, not its multiple child attributes which are separately a map that may or may not have values. If you look at `readCloudHsm2ClusterCertificates()` you are only ever assigning the first element of return slice (`return []map[string]interface{}{certs}`), not appending multiple maps to that slice. The cluster only potentially has one `aws_hardware_certificate`, for example.

Its not reliable to reference `TypeSet` attributes in other portions of Terraform configuration since its `Set` hash will change depending on its contents. e.g. to read `aws_hardware_certificate` from a `TypeSet`, the reference will become `${aws_cloudhsm_v2_cluster.example.cluster_certificates.######.aws_hardware_certificate}` where `######` will be an integer determined by hashing the values of `cluster_certificate`, `cluster_csr`, `aws_hardware_certificate`, `hsm_certificate`, and `manufacturer_hardware_certificate`. Operators won't be able to calculate that value (easily) on their own. Making the parent `TypeList` makes that reference easily `${aws_cloudhsm_v2_cluster.example.cluster_certificates.0.aws_hardware_certificate}`.

To put all this another way, it might be clearer to see this if it was instead something that could be configured in HCL. By fake example, you could never do:

```hcl
resource "aws_cloudhsm_v2_cluster" "example" {
  cluster_certificates {
    cluster_certificate = ""
  }

  cluster_certificates {
    cluster_certificate = ""
  }
}
```

So its unnecessary to worry about whether `cluster_certificates` itself is a list or set since it will only have one element itself. When we have a list/set that is only one element, we prefer single element list `parent.0.child` for easier reference.